### PR TITLE
Manually created dynamic links should be subject to allowed/blocked check

### PR DESCRIPTION
**Pull request by palexsua**
- **Description**: Dynamic links can be validated against a block/allow list, which is defined server side in the Firebase Console. This verification should apply to both short and long dynamic links, and be supported by both iOS and Android SDKs.
- **What’s expected**: It’s expected that both iOS and Android will verify both kinds of dynamic link, and if it passes the verification rules defined server side, allow the link to be processed.
- **What’s observed**: On iOS, long dynamic links are not verified, and it’s possible for a link that doesn’t pass the "allowed link" rules to be treated as verified. Short dynamic links are verified, that’s not an issue. This also isn’t an issue on Android, which verifies both short and long dynamic links.
- **Proposed solution**: Ensure the error response from `reopenAttribution` (which contains information to whether the link was verified or not), should be propagated further up for long dynamic links, so the appropriate decision on whether it’s a verified link can be made. Some method definitions have changed as well to accommodate this, which have now made them asynchronous.

According to the source code all short links are checked when a call to `FIRDynamicLinkNetworking` class method `resolveShortLink:completion:` takes place. This method does a HTTP call to `reopenAttribution` endpoint and receives a response with an error, if the URL is not allowed. The problem is that the same is not true for long dynamic links (aka [manually created links](https://firebase.google.com/docs/dynamic-links/create-manually)) - no call to `resolveShortLink:completion:` method (or its long links counterpart) is done. Thus all URLs are accepted and considered allowed.
Another part of the problem is that when a call to `reopenAttribution` endpoint is done the response with the error description is ignored and and error equal to `nil` is passed to completion callback. Thus, all traces of the issue (the URL being not allowed) are lost.

### CODE REVIEW

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:119]: Whenever a call to `reopenAttribution` endpoint took place response object was disregarded despite the fact that its status code indicated an error. On the other hand, `error` object itself was not populated and was `nil`. As a result, all traces of an issue used to be lost here and the calling code had no clue that the url was not allowed.

Screenshot 2020-04-29 at 17 58 06

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:74]: Whenever a call to `reopenAttribution` endpoint took place response object was disregarded despite the fact that its status code indicated an error. On the other hand, `error` object itself was `nil`. As a result, all traces of an issue used to be lost here and the calling code had no clue that the url was not allowed.

Screenshot 2020-04-29 at 17 58 06

**Comment by paulb777** [comment]
@palexsua Thanks for the comprehensive PR!  We'll do our best to get a review soon, although it may be a challenge in the very near term with WWDC next week.

One initial question: Is there a way to make the asynchronous API additive instead of breaking? Changing the current API will require holding off on merging this until the next major release.

**Comment by palexsua** [comment]
> @palexsua Thanks for the comprehensive PR! We'll do our best to get a review soon, although it may be a challenge in the very near term with WWDC next week.
> 
> One initial question: Is there a way to make the asynchronous API additive instead of breaking? Changing the current API will require holding off on merging this until the next major release.

@paulb777 Thank you! I think I can make the asynchronous API additive. Do you want me to open a new PR or update the existing?

**Comment by palexsua** [comment]
@paulb777 @eldhosembabu any update on these changes?

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (ncooke3) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:130]: Curious, what is the context here in deciding which error to pass in the completion? 

Also, nit: perhaps rename `err` to `responseTypeError`, `customError`, or something like that

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (ncooke3) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:123]: nit: maybe rename to something like `completedResponse`

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (ncooke3) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:132]: If I'm not mistaken, `resp` is only used within this `else`'s scope, correct? Perhaps its declaration could be moved  to within the `else` block.

**Comment by palexsua** [comment]
@ncooke3 Thanks for the comments! I'll have a look later today. I've also pushed one more commit as @paulb777 requested for additive changes.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:130]: Per `FIRNetworkRequestCompletionHandler` definition `error` object is nullable. I decided to give precedence to the original error over the local one (`err`), because I didn't want to shadow potentially more serious issue, e.g. stemming from server side, given that initially `response` object had been disregarded completely.

Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:130]: Renamed `err` -> `responseTypeError`.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:132]: Moved.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:123]: As we're primarily interested in `statusCode` I decided to get rid of `resp` object altogether: `NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];`.

**Comment by palexsua** [comment]
@ncooke3 Addressed your review comments in https://github.com/firebase/firebase-ios-sdk/pull/5853/commits/0c20e3bb6148a8dbd24a4712117ffbd2f4e14c93. Thanks!

**Comment by palexsua** [comment]
@ryanwilson I guess after I've made the changes additive `public-api-change` label is not relevant anymore, isn't it?

**Comment by paulb777** [comment]
@palexsua Thanks for refactoring. It's still an API change. It's now no longer an API break so we can potentially merge this for a minor update release instead of waiting for the next rarer major release.

The API change will still need an internal API review. cc: @christibbs, @ncooke3

**Comment by palexsua** [comment]
@paulb777 Got it. Thanks!

File: FirebaseDynamicLinks/Sources/Public/FIRDynamicLinks.h
Comment (ncooke3) [inline FirebaseDynamicLinks/Sources/Public/FIRDynamicLinks.h:94]: @ryanwilson @paulb777 Are there any implications to adding a deprecated message to a currently public API?

Comment (paulb777) [reply FirebaseDynamicLinks/Sources/Public/FIRDynamicLinks.h:94]: @ncooke3 It should be fine if the API council agrees with the transition recommendation.

**Comment by christibbs** [comment]
> @palexsua Thanks for refactoring. It's still an API change. It's now no longer an API break so we can potentially merge this for a minor update release instead of waiting for the next rarer major release.
> 
> The API change will still need an internal API review. cc: @christibbs, @ncooke3

Thanks for the work @palexsua. I'll kick off the internal API review with @eldhosembabu.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:131]: "responseTypeError" is used only if error is nil; so can we defer the responseTypeError error creation?

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:127]: The whole method change that I can see is about constructing the error object and passing it through. Can we refactor this with a new method which can take in the NSHTTPURLResponse object and returns an NSError object and move all your code changes there?

Then in this method you can call the new method without have much changes to the existing method. This would make the code more readable and we can test the new method individually. Also avoids duplicated lines of code to call handler(url,error) from multiple if else blocks.

So the new method would be something like (Feel free to change the name and signature of the method) : 

   `+(nullable NSError *) extractErrorFromResponse:(NSURLResponse *)response
                            withError:(nullable NSError *) error{
  if(error){
    return error;
  }

  NSError *customError = nil;

  //Your logic to create the custom error.

  return customError;
}`

In current completion handler, we will just call : 

`error = [FIRDynamicLinkNetworking extractErrorFromResponse:response
             withError:error];`

and the rest of the code will remains same. Let me know if that makes sense.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:493]: could you please explain why you think we need to check for the existence of "dynamicLinkFromUniversalLinkURL:completion" here?

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:443]: dynamicLinkFromUniversalLinkURL:completion and dynamicLinkFromUniversalLinkURL both shares the same code apart from completion block calling part. So can we refactor the common code to a method which can be called from dynamicLinkFromUniversalLinkURL:completion and dynamicLinkFromUniversalLinkURL which would avoid code duplication and would make the code maintainable.

**Comment by eldhosembabu** [comment]
hi @palexsua sorry for reviewing the code late from my side. I've left a few code review comments. Please let me know if you have any questions .

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:131]: Sure. Fixed.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:127]: I got the point.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:493]: Well, initially I introduced it for compatibility purposes, but now I see that it's not needed.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:443]: `completion` callback is called in multiple (2) places inside `dynamicLinkFromUniversalLinkURL:completion` method. Moreover, these methods differ in return value - `void` vs `FIRDynamicLink`. The idea is to deprecate `dynamicLinkFromUniversalLinkURL:` now and remove completely in the upcoming releases. Do you still think it makes sense to refactor it? If you insist though, I can do it. No problem.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:127]: Please, see the implementation and if it's ok with you I can add a couple of unit tests.

**Comment by palexsua** [comment]
@eldhosembabu Thank you for your review comments! I appreciate. I've pushed the changes and replied to your comments. Please, have a look.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:122]: Don't we need to check data before using?

Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m]: Right. Added the check here: https://github.com/firebase/firebase-ios-sdk/pull/5853/commits/73f3bb291fbd0b37c7e99fe95033c58e0c03fbd2

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:443]: I agree with the fact that the other function is getting deprecated and will get eventually removed but we are not sure when we are going to do it. Since removing the other method would be a breaking change, I'm not sure that we can do it sooner. So until then, if we need to make any fixes in dynamicLinkFromUniversalLinkURL logic, we will end up having it in two places.

My thought was to have a common internal method which can take the completion handler as well returns the dynamic link so that when called from `dynamicLinkFromUniversalLinkURL:` we will return based on the return of this internal method otherwise if called from `dynamicLinkFromUniversalLinkURL:completion`,we will ignore the return and rely on the completion block. So my proposal would be something like: 

    (nullable FIRDynamicLink *)dynamicLinkInternalFromUniversalLinkURL:(NSURL *)url
                                  completion:(FIRDynamicLinkUniversalLinkHandler)completion {
     //Have existing logic here and make call to completion block only if its not nil.
     ....
     if(completion){
     completion(dynamicLink,error);
     }
     ....

When calling this internal function from `dynamicLinkFromUniversalLinkURL:` just do :

    return  [self dynamicLinkInternalFromUniversalLinkURL:url completion:nil]

  And when calling from `dynamicLinkFromUniversalLinkURL:completion` just call : 

    [self dynamicLinkInternalFromUniversalLinkURL:url completion:completion]

I'm not sure this is a good coding pattern but this would help us with code repetition. Once we remove the old method completely, we can get rid of this internal method. Since my experience with Obj C is limited, I would prefer to get advice on this from @paulb777

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (eldhosembabu) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m]: seems good to me apart from a minor comment regarding checking the variable.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m]: Thanks for removing.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m:131]: Can you move this data check to parent `else if` so that you don't need to repeat 
     
    customError = [FIRDynamicLinkNetworking genericErrorForUrl:url];

Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m]: Addressed in https://github.com/firebase/firebase-ios-sdk/pull/5853/commits/11e614d0292f1369947f2e65e4cf4d4873a5ff65.

**Comment by christibbs** [comment]
Update from my end: we've kicked off the internal Firebase API review process.

**Comment by palexsua** [comment]
@eldhosembabu @christibbs any update? please, let me know, if any changes are required.

**Comment by eldhosembabu** [comment]
hi @palexsua now we are on our way towards the api review where the new proposed api needs to be approved by the api council...it would take O(weeks) time to get the approval..If we are not getting followup questions, we should get it approved in a week or two. After that we should be able to merge this PR if its all good.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (paulb777) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:443]: Agreed. It's worth refactoring to a common internal method - even if it's only a few months - since this much duplicate code could easily lead to inconsistent behavior or bugs.

**Comment by palexsua** [comment]
@eldhosembabu Got it. Thank you for the update!

**Comment by palexsua** [comment]
@eldhosembabu @christibbs guys, any update?

**Comment by christibbs** [comment]
> @eldhosembabu @christibbs guys, any update?

We're working on the API change approval right now. Thanks for your patience.

**Comment by palexsua** [comment]
@eldhosembabu @christibbs guys, any update?

**Comment by palexsua** [comment]
@eldhosembabu @christibbs guys, any update?

**Comment by palexs** [comment]
@paulb777 @eldhosembabu @ncooke3 guys, let's merge this PR.

**Comment by sbaruth** [comment]
Would also love to see this get merged, if it's ready to go.

**Comment by christibbs** [comment]
Hey @palexsua @sbaruth. Really sorry for the delay. We will prioritize this and get through API review this week.

**Comment by palexs** [comment]
@christibbs Thanks for taking care of it! Looking forward to hearing the API review verdict ;)

**Comment by christibbs** [comment]
One additional thing: could you provide an example of if/how you're using the new `dynamicLinkFromUniversalLinkURL:completion:` method in your app?

Is it similar to #6 [here](https://firebase.google.com/docs/dynamic-links/ios/receive)? Returning synchronously from `application:continueUserActivity:restorationHandler:` but using the completion for a side effect?

**Comment by palexsua** [comment]
@christibbs Actually, we do not call `dynamicLinkFromUniversalLinkURL:completion:` directly in our app, but we rely on code that uses it. The issue I observed was that long dynamic links bypassed any sort of verification and when I traced back the issue I ended up seeing that response from `reopenAttribution` endpoint was not being "waited for". So, the gist of the proposed fix is to wait for the response and pass allowed/blocked status via `completion` callback.

**Comment by christibbs** [comment]
@palexsua We got the API approval! @eldhosembabu had one outstanding issue with the implementation, but we can merge once that's resolved.

**Comment by palexsua** [comment]
@christibbs That's great news! Thanks for keeping me in the loop.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:433]: @palexsua could you please address this prior to merging?

Comment (palexsua) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:445]: @eldhosembabu Addressed in https://github.com/firebase/firebase-ios-sdk/pull/5853/commits/0944814ddf4b11cf59f258439797ce3e8dd965a5

**Comment by palexsua** [comment]
@christibbs I've addressed the change suggested by @eldhosembabu.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:445]: thanks @palexsua

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (eldhosembabu) [inline FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:435]: I'm not sure whether this completion block calling is useful here since it would just invoke the block without giving any clue on which url we have tried and failed and so i believe this can be removed. thoughts @christibbs @paulb777 ?

**Review by eldhosembabu — COMMENTED** [review]
Looks good to me.

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (paulb777) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:435]: It may be needed to satisfy the API if it expects there to be a callback.

However, as this is written, there could be two callbacks if the nested if is reached above and that doesn't seem right.

**Review by paulb777 — CHANGES_REQUESTED** [review]
Thanks for your perseverance on this nice API improvement!

In addition to resolving the open callback question, please update https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseDynamicLinks/CHANGELOG.md with a description of the deprecation and new API. (The next planned release version is v7.3.0)

File: FirebaseDynamicLinks/Sources/FIRDynamicLinks.m
Comment (christibbs) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:435]: @paulb777 We return on line 430, which should prevent 422 and 436 from both being fired.

Comment (eldhosembabu) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:435]: I agree with you @paulb777; but i think the 3p users can't do anything with the callback as they don't know to the minimum what url we have tried to parse.

I could be wrong but I'm not seeing a chance for two callback method invocation since we are returning right after the service call in line 425.

Comment (paulb777) [reply FirebaseDynamicLinks/Sources/FIRDynamicLinks.m:435]: Oh right, agreed that there can't be two callbacks.  I missed that return.

On the other question, couldn't 3p users check for the nil parameters to know that the url failed to parse or that `url.query.length` was 0?

**Comment by palexsua** [comment]
@paulb777 Afaiu, the open callback question is resolved, isn't it? So, now it's only a matter of updating the changelog?

**Comment by paulb777** [comment]
@palexsua Agreed. Sorry about the confusion. I'll merge it here and we can follow up on the changelog separately.

**Review by paulb777 — APPROVED** [review]

**Comment by christibbs** [comment]
@palexsua Thanks for your patience. And congrats: now you are a Firebase contributor!

**Comment by palexsua** [comment]
@christibbs Thank you! It feels great! 🥇

**Pull request merged**
